src/mem/slab.h

```diff
@@ -163,27 +159,19 @@ namespace snmalloc
         error("Detected potential double free.");
 #endif
       meta.debug_slab_invariant(is_short(), this);
-      meta.sub_use();
-
-      bool was_full = meta.is_full();
-      if (unlikely(was_full))
-        return false;
 
-      bool is_unused = meta.is_unused();
-      if (unlikely(is_unused))
+      if (unlikely(meta.return_object()))
         return false;
 
       // Update the head and the next pointer in the free list.
-      uint16_t head = meta.head;
-      uint16_t current = pointer_to_index(p);
+      void* head = meta.head;
 
       // Set the head to the memory being deallocated.
-      meta.head = current;
-      assert(meta.valid_head(is_short()));
+      meta.head = p;
+      assert(meta.valid_head());
 
       // Set the next pointer to the previous head.
-      Metaslab::store_next(
-        p, (head == 1) ? nullptr : pointer_offset(this, head));
+      Metaslab::store_next(p, head);
       meta.debug_slab_invariant(is_short(), this);
```

Comment:
Does this elide `is_short()` in the NDEBUG case when compiled? Looks like it should, just double checking.

Reply:
Pretty hard to tell given all the inlining. I have rewritten to push the is_short() inside the function, and thus inside the ifndef.